---
output:
  pdf_document:
    fig_height: 3
    latex_engine: pdflatex
header-includes:
   - \usepackage{helvet}
   - \renewcommand*\familydefault{\sfdefault}
   - \usepackage[T1]{fontenc}

geometry: margin=0.5in
---

```{r setup, include=FALSE}

knitr::opts_chunk$set(echo = TRUE)


library(data.table)
dict <- fread(DICT_FILE,data.table=F, check.names = T)
rownames(dict) <- dict$Methylation.Class.Name.Abbreviated

votes$methClass <- dict[rownames(votes),]$Methylation.Class.Name

mcf <- votes %>%
 left_join(dict, by=c("methClass" = "Methylation.Class.Name")) %>%
 group_by(Methylation.Class.Family) %>%
 summarise(cal_Freq_MCF = sum(cal_Freq))
```

```{r, out.height= "50px", echo=FALSE, fig.align="right"}
knitr::include_graphics("header_PDFreport.pdf")
```


## Nanopore sequencing report

# `r snakemake@wildcards[["sample"]]`

## Quality control metrics

#### Barcode statistics

```{r comment='', echo=FALSE}
cat(readLines(DEMUX), sep = '\n')
```

#### Read statistics

```{r comment='', echo=FALSE}
cat(readLines(NANOSTAT, n=15), sep = '\n') # Read general summary (first 15 lines) of NanoStat statistics
```

Mean genome coverage is `r meanDepth`X.

```{r, out.height= "150px", echo=FALSE}
knitr::include_graphics(RL)
knitr::include_graphics(GC)
```

## Copy number profile

```{r, out.height = "180px", echo=FALSE}
knitr::include_graphics(CN)
```

## Methylation-based classification

Methylation-based classification is based on **`r as.integer(model_info$num.features)`** CpG sites (overlap of sites covered in this sample and the training set).
The out-of-bag error of this ad-hoc classifier is **`r round(model_info$oob.error*100, digits=1)`%**.
Using this classifier, the sample has been classified as **`r ifelse(max(votes$cal_Freq)>=0.15, dict[rownames(votes[which.max(votes$cal_Freq),]),"Methylation.Class.Name"], "not classifiable (score < 0.15)")`**.
This prediction has a confidence score of **`r round(max(votes$cal_Freq), digits=3)`**.
At the methylation class **family** (MCF) level, the sample has been classified as **`r ifelse(max(mcf$cal_Freq_MCF)>=0.15, mcf[which.max(mcf$cal_Freq_MCF),]$Methylation.Class.Family, "not classifiable (score < 0.15)")`**. The MCF prediction has a confidence score of **`r round(max(mcf$cal_Freq_MCF), digits=3)`**.

The raw and calibrated scores for the Top 10 entities are given below.
Vertical dashed lines indicate the recommended >0.15 cut-off (Kuschel et al., Neuropathol Appl Neurobiol 2022).

```{r, out.height = "200px", echo=FALSE, warning=FALSE}
library(ggplot2)

ggplot(data=votes, aes(x=methClass, y=Freq)) +
  geom_bar(stat="identity", fill="#1f78b4") +
  coord_flip() +
  scale_x_discrete(limits=votes[order(votes$Freq,decreasing=T),]$methClass[10:1]) +
  ylab("votes (%)") + xlab("methylation class") +
  theme_classic() + theme(aspect.ratio=1)
```

```{r, out.height = "200px", echo=FALSE, warning=FALSE}
ggplot(data=votes, aes(x=methClass, y=cal_Freq)) +
  geom_bar(stat="identity", fill="#dc143c") +
  geom_hline(yintercept=0.15, linetype="dashed") + # indicate cut-off as in Kuschel et al., NAN 2022
  coord_flip() +
  scale_x_discrete(limits=votes[order(votes$cal_Freq,decreasing=T),]$methClass[10:1]) +
  ylim(0,1) +
  ylab("confidence score") + xlab("methylation class") +
  theme_classic() + theme(aspect.ratio=1)
```

```{r, out.height = "200px", echo=FALSE, warning=FALSE}
ggplot(mcf, aes(x=Methylation.Class.Family, y=cal_Freq_MCF)) +
  geom_bar(stat="identity", fill="#dc143c") +
  geom_hline(yintercept=0.15, linetype="dashed") + # indicate cut-off as in Kuschel et al., NAN 2022
  coord_flip() +
  scale_x_discrete(limits=mcf[order(mcf$cal_Freq_MCF,decreasing=T),]$Methylation.Class.Family[10:1]) +
  ylim(0,1) +
  ylab("confidence score") + xlab("methylation class family") +
  theme_classic() + theme(aspect.ratio=1)
```

### CUDA TSNE plot

Computed with the same hyperparameters as the "standard" TSNE.

```{r, echo=FALSE}
knitr::include_graphics(cudaTSNE)
```

#### Disclaimer

Methylation-based classification using nanopore whole genome sequening is a research tool currently under development. It has not been clinically validated in sufficiently large cohorts. Interpretation and implementation of the results in a clinical setting is in the sole responsibility of the treating physician.



(Report generated on `r Sys.time()`. Pipeline version: `r system("git describe --tags", intern = T)`, `r system("git log --pretty=format:'%h' -n 1", intern=T)`)
